### ISSUE TITLE
Narrowing types fail when either types can be None

### ISSUE THREAD

**Issue by Bluenix2** [microsoft/pyright#2219]
**To reproduce**
I have the following code (simplified and reproduced), which demonstrates the problem:
```py
def either(a: Optional[int], b: Optional[int]) -> int:
    if a is None and b is None:
        raise TypeError("both 'a' and 'b' is None")

    return a if a is not None else b
```
```
Expression of type "int | None" cannot be assigned to return type "int"
  Type "int | None" cannot be assigned to type "int"
    Type "None" cannot be assigned to type "int"
```

**Expected behavior**
I would expect this to work, because the if-statement filters out the case where both are None.
Which means that by the return only one of them should be None.

**VS Code extension or command-line**
![image](https://user-images.githubusercontent.com/38372706/130472064-810e3e45-6b21-42eb-a0bc-c8ab74205301.png)

**Comment by erictraut** [comment]
This is expected behavior. Type narrowing is done on individual expressions. A type that is narrowed for one expression cannot depend on the type of another expression.

In your sample above, there are two expressions involved: `a` and `b`. There is an implied `else` condition which is the inverse of the `if` condition: `a is None or b is None`. Neither `a` nor `b` can be narrowed in this code path. They are both still of type `Optional[int]`.

You might think that a type checker would be able to determine that the type of `a` is `int` _if the type of `b` is `None`_ but `Optional[int]` if the type of `b` is `not None`. Tracking conditional types like that isn't feasible within a static type checker. It would involve substantial complexity, computation, and memory overhead and would explode exponentially in the general case. No static type checkers of compilers that I'm aware of (in any language) perform this type of conditional type tracking.

You can avoid this situation through a simple change to the code:
```python
def either(a: Optional[int], b: Optional[int]) -> int:
    if a is not None:
        return a

    if b is None:
        raise TypeError("both 'a' and 'b' is None")

    return b
```

Or if you prefer...

```python
def either(a: Optional[int], b: Optional[int]) -> int:
    result = a if a is not None else b
    if result is None:
        raise TypeError("both 'a' and 'b' is None")
    return result
```

**Comment by Bluenix2** [comment]
Thank you very much for the explanation, I understand. It would be extremely complicated to keep track of "`x` is `A` if `y` is `B` and `z` is `C` if `w` is `A`, etc."

Even though we have narrowed down the cases to *one of them being `Optional[X]`*, they're still individually `Optional[X]`.
